docs/adrs/003-store-room-allocation-in-memory.md

```diff
@@ -0,0 +1,84 @@
+---
+status: accepted
+date: 2024-05-21
+deciders: Kamil Kołodziej, Jakub Pisarek, Radosław Szuma
+consulted: Michał Śledź
+informed: Fishjam Team
+---
+# Route incoming traffic within the cluster without a need to provide room_id
+
+## Context and Problem Statement
+
+In the current architecture after we create a room in cluster we return on which instance (Jellyfish host) it was created.
+This means that Jellyfish client needs to make request to only that specific instance.
+Which effectively leads to the situation where client may need to keep mapping between room_id and jellyfish host.
+
+We would like to hide information about internal architecture of the cluster and the exact JF hosts as well as make it easier for clients so they can use only cluster endpoint.
+
+## Decision Drivers
+
+* Cluster architecture should be an internal information
+* `room_id` should be a sufficient information to route the requests
+
+## Considered Options
+
+* Store room allocation across the cluster using in-memory ETS
+* Store room allocation across the cluster using database
+* Leave it as it is and update the SKDs
+
+## Decision Outcome
+
+Chosen option: Store room allocation across the cluster using in-memory ETS.
+
+### Consequences
+
+- Routing within the cluster will take some of the responsibility from the client.
+- This responsibility is moved to the internal part of the Fishjam.
+- We may encounter some issues with data not being propagated quick enough or some message missing and will need to handle those scenarios.
+- We need to add some observability tools/metrics to allow us detect any issue within the cluster.
+- When it comes to the websocket we would prefer to forward it to the correct node instead of returning exact ws address or forward traffic.
+
+We want to try it first with the data stored in ETS, because since our tool is an open source one, we don't want to enforce users to add database.
+If this solution turns out not feasible, we will move towards using some external database or reiterate on ours research.
+
+## Pros and Cons of the Options
+
+### Store room allocation across the cluster using in-memory ETS
+
+This option assumes that every node will keep track of the room allocation across the cluster in ETS.
+The exact implementation may be different but the general idea is to publish to other nodes rooms on the local node.
+Based on that factor it doesn't matter which node will receive the request we can guide it to the correct node.
+
+* Good, because we drop the requirement for client to make request to specific instance
+* Good, because it's a step towards the deployment without downtime
+* Good, because we are in control of routing inside the cluster and internals do not leak out
+* Bad, because if the data is eventually consistent and if we make 2 requests very quick (create room, add peer) second one can fail.
+    This is because second request may land in a node which don't have information about that room yet.
```

Comment:
**comment**: I imagine that an FJ instance behind a load balancer would be a pretty common setup. I think this exact scenario should be the main focus when handling issues with unpropagated clustering information.

Reply:
Well this one could be handled in two ways:
- introduce central storage like database which will determine how to route the traffic so this issue won't be happening
- cipher/hash/sign the room_id in a way that it has the necessary information for every node to determine the owner of the room only by room_id

Reply:
We could also, when we dont have a room id in our cache, ask the cluster. 

I assmue that when we broadcast something on Phoenix PubSub we dont know when subscribers will receive our message?